src/helpers/visualisation.py

```diff
@@ -428,4 +428,715 @@ def plot_differences_for_group(
+    """
+    Convert list of label sequences to Sankey node/link arrays.
+    flows: [[l0,l1,l2,...], ...] (user-turn labels only)
+    Returns: (node_labels, source_idx, target_idx, values)
+    """
+    from collections import Counter as _Counter
+
+    stage_labels = [set() for _ in range(max_user_turns)]
+    links_counter = _Counter()
+
+    for seq in flows:
+        seq = seq[:max_user_turns]
+        if len(seq) < 2:
+            # still register single nodes so they appear
+            for i, lbl in enumerate(seq):
+                stage_labels[i].add(lbl)
+            continue
+        for i, lbl in enumerate(seq):
+            stage_labels[i].add(lbl)
+        for i in range(len(seq) - 1):
+            links_counter[(i, seq[i], seq[i+1])] += 1
+
+    node_labels, node_index = [], {}
+    for s in range(max_user_turns):
+        ordered = sorted(stage_labels[s]) if sort_labels_within_stage else list(stage_labels[s])
+        for lbl in ordered:
+            node_index[(s, lbl)] = len(node_labels)
+            suffix = f" {stage_label_func(s)}" if stage_label_func else ""
+            node_labels.append(f"{lbl}{suffix}")
+
+    src, tgt, val = [], [], []
+    for (s, a, b), cnt in links_counter.items():
+        if (s, a) in node_index and (s+1, b) in node_index:
+            src.append(node_index[(s, a)])
+            tgt.append(node_index[(s+1, b)])
+            val.append(cnt)
+    return node_labels, src, tgt, val
+
+
+def plot_sankey(
+    flows: List[List[str]],
+    max_user_turns: int = 4,
+    stage_label_func: Optional[Callable[[int], str]] = lambda s: f"(t{2*s})",
+    title: str = "",
+    width: int = 980,
+    height: int = 560,
+    font_size: int = 12,
+    save_html: Optional[str] = None,
+    show: bool = True, 
+
+
+):
+    import plotly.graph_objects as go
+    labels, s, t, v = sankey_from_flows(
+        flows, max_user_turns=max_user_turns, stage_label_func=stage_label_func
+    )
+    fig = go.Figure(data=[go.Sankey(
+        arrangement="snap",
+        node=dict(pad=18, thickness=18, line=dict(color="black", width=0.5), label=labels),
+        link=dict(source=s, target=t, value=v)
+    )])
+    fig.update_layout(title=title, font_size=font_size, width=width, height=height)
+    if save_html:
+        os.makedirs(os.path.dirname(save_html), exist_ok=True)
+        fig.write_html(save_html)
+    if show:
+        fig.show()
+    return fig
+
+# Build flows from tidy per-turn DF
+def build_flows_from_tidy(
+    df: pd.DataFrame,
+    id_col: str,
+    turn_col: str,
+    label_col: str,
+    sort_by_cols: Optional[Sequence[str]] = None,  
+) -> List[List[str]]:
+    data = df.copy()
+    if sort_by_cols:
+        data = data.sort_values(list(sort_by_cols))
+    flows = []
+    for _, sub in data.groupby(id_col, sort=False):
+        sub = sub.sort_values(turn_col)
+        flows.append(sub[label_col].tolist())
+    return flows
+
+
+
+
+
+
+
+
+def build_transition_matrix_from_tidy(
```

Comment:
For any functions that do not explicitly plot something, can you create a new file in src/helpers/multiturn_analysis_helpers.py and put them all in there. visualization.py should just be functions that create figures ideally, with no specific logic beyond the figure creation